### DISCUSSION TITLE
How would you adapt SHAP for a multivariate time series model?

### DISCUSSION THREAD

**Discussion by loewenm** [shap/shap#1997] [Q&A]
I'm trying (and failing) to adapt SHAP to work with a multivariate time series model.

The issue arises due to SHAP requiring (being designed for) input shape [samples, features]. However, time series data has an extra dimension with shape [samples, timesteps, features].

I was hoping that there'd be some interesting way to wrap a model that takes data as a 2D input shape, and transforms it into the desired 3D shape to feed into models like LSTM, etc. Then I would call SHAP on the wrapper model... But would this defeat the purpose of what's working under SHAP's hood?

I wonder if you could simply _flatten_ the dataset to [samples, timesteps * features], but this wouldn't work in a typical time series-oriented model either...

For reference, my code is:

```
# select a set of background examples to take an expectation over
background = X_pred[np.random.choice(X_pred.shape[0], 21, replace=False)]
# shape = (21, 252, 16)

explainer = shap.KernelExplainer(model.predict, background, link="logit")

shap_values = explainer.shap_values(X_pred[-30:])
```

Which, of course, results in the following error:

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-46-208fa6d6bdbf> in <module>()
      1 # explaining each prediction requires 2 * background dataset size runs
----> 2 shap_values = explainer.shap_values(X_pred[-30:])

/usr/local/lib/python3.7/dist-packages/shap/explainers/_kernel.py in shap_values(self, X, **kwargs)
    154             X = X.tolil()
    155         assert x_type.endswith(arr_type) or sp.sparse.isspmatrix_lil(X), "Unknown instance type: " + x_type
--> 156         assert len(X.shape) == 1 or len(X.shape) == 2, "Instance must have 1 or 2 dimensions!"
    157 
    158         # single instance

AssertionError: Instance must have 1 or 2 dimensions!
```

**Has anyone seen an example of a working multivariate time series model using SHAP to extract feature importance???**

Any help would be greatly appreciated.

**Comment by loewenm** [comment]
Follow-up question: Are there any _Explainable AI_ libraries available out there (not necessarily SHAP) that have time series capabilities built in?

**Reply by XiangBu** [reply]
hey，you can try this, see if it can solve ur problem.  https://www.kaggle.com/ltsmssv/shap-for-lstm-hpccv2

**Comment by St-Savin** [comment]
I did use SHAP on a GRU model.  I had use "tf.compat.v1.disable_v2_behavior()" since there was some problems with the version of tf.  Also, I don't know if what I did was good, but I managed to shape the data and shap value arrays so they can be processed by shap.summary_plot .

**Reply by alexxony** [reply]
can you show me whole code

**Reply by pedro-zenario** [reply]
Can you please show me the code?

**Reply by St-Savin** [reply]
The trick is to reshape the 3D array of the Shapvalue in 2D for the time interval that you want. 

For my data, I had a 24 parameters model with a 5 timesteps model, the shapvalues were calculated on 99 slices of 5  x 24 datapoints.

# First calculate the shapevalues
shap_values = explainer.shap_values(test_x[1:100])

# Create the arrays for the shape values and the data

tempo = np.array(shap_values)
dates = test_x[1:100]
tempo = np.squeeze(tempo)

#Then use those arrays to extract the shape values for the interval of timesteps that you want
# here I want just the shap values for the timestep #1

# first time
tempsdep = 1
# last time
tempsfinal = 1
# number of slices
tempsshap = 99

m=0
Shapvalues_mod = []
Data_mod = []

i = tempsdep
j = 0
k = 0
for x in range(tempsshap*(tempsfinal-tempsdep + 1)):
    Shapvalues_mod.append(tempo[k,i,:])
    Data_mod.append(dates[k,i,:])
    i = i + 1
    j = j + 1
    if i > tempsfinal :
        i = tempsdep
        k = k + 1

Shapvalues_mod = np.array(Shapvalues_mod)
Data_mod = np.array(Data_mod)

# finally create the dateframe for the shape values use the shapvalue as if it was a 2d array
my_list = list(datamod)

Dftest = pd.DataFrame(data = Data_mod
             , columns = [my_list])

shap.summary_plot(Shapvalues_mod, Dftest)

**Reply by pedro-zenario** [reply]
The problem is that I can't even run succesfully explainer.shap_values(...). Can you please show me your code from the beginning/explain it to me so that I can try and solve my situation?

**Reply by Jeetd7** [reply]
Is this problem is resolved , i am not able to apply shap on gru model time series data

**Reply by Jeetd7** [reply]
@pedro-zenario You able to do it now? or still having the same issues

**Comment by mihailyanchev** [comment]
Hi, it seems that using GradientExplainer helps deal with 3d shapes. However, if you use tf 2.x you would probably run into the issue with TypeError: 'NoneType' object cannot be interpreted as an integer described [here](https://github.com/slundberg/shap/issues/1551) 

I was hoping that 0.40.0 fixes that, but apparently no.

**Reply by jintiandefanzhenhaochi** [reply]
Yes, and the GradientExplainer method seems to have limitations, requiring that the output must be a single tensor, but my network is multi-class, and if argmax is added in the middle to convert to one-dimensional, then there will be a case where the gradient cannot be passed T _ T

**Comment by St-Savin** [comment]
What kind of explainer did you use ?  This line worked a 2 years ago. Maybe it does not work now.  

explainer = shap.DeepExplainer(regressorGRU, train_x[1:1000])

**Comment by jintiandefanzhenhaochi** [comment]
Hello, I encountered the same problem. I checked the library file, if using KernelExplainer, then the input dimension must be one-dimensional or two-dimensional, but with DeepExplainer, there is no such limit, but I try to use the latter process, because my output is two-dimensional, so there are other problems. 
At present, I am trying to use KernelExplainer. The input data is DataFrame framework, which requires two-dimensional data. I extracted one layer of the third dimension and changed it to two-dimensional data, but the effect of the network is very poor. Therefore, I am also considering the two-dimensional input you proposed, and then transferred to the LSTM layer in the middle of the three-dimensional. 
I suggest you try DeepExplainer first, maybe your code is suitable for it. If I have a new solution, I will reply again.

**Reply by nickparative** [reply]
For anyone looking for a solution

Currently the last push for this bug in the shap package was Jan 1st, so not much progress has been made.

I did find a wrapper though that does the job: https://github.com/vsubbian/WindowSHAP

You just add the windowshap.py file to your directory and import the functions. The benefit here is that there is even a function for sliding window data. Convert your tensors to numpy, using the training data as the background_data and the test_data for local shap calculations

`
### input data shape: (batch, sequence, features)

### shuffle and retrieve data from the training tensor
x_train_np = x_sequence_train[:1000].numpy()
np.random.shuffle(x_train_np)
background_data = x_train_np
test_data = x_sequence_test[:1].numpy()

model = reloaded_model

### Initialize SHAP explainer
ts_phi_2 = np.zeros((len(test_data),test_data.shape[1], test_data.shape[2]))
for i in range(len(test_data)):
    window_len = 5 # This is the number of timepoints in your sequence
    stride = 1 # this is your sliding window stride
    gtw = windowshap.SlidingWindowSHAP(model, stride, window_len, background_data, test_data[i:i+1], model_type='lstm')
    ts_phi_2[i,:,:] = gtw.shap_values()
`
The output is a numpy array in the shape of your test_data. In my case that was (1,5,12) (1 set of sequences, 5 sequences, 12 features)

Here I sum all the values for my data and then I have a list of 12 feature names that I use to create my dataframe.
`
shap_values_total = np.sum(ts_phi_2, axis=1)

shap_df = pd.DataFrame(shap_values_total, columns=flattened_feature_names)
shap_df.sum(axis=0).astype(float).round(decimals=3)
`

There's also an argument for these functions to accommodate models with branching inputs--specifically time series data and static data. 

If anyone comes up with ways to generate global versus local values, please post an explanation. Thanks!

**Reply by matt-buckley** [reply]
@jintiandefanzhenhaochi the DeepExplainer documentation seems to indicate that it only works with scalar outputs: "SHAP values are specific to a single output value, so the output tf.Tensor should be a single dimensional output (,1)." I haven't been able to get DeepExplainer to work, would be interested to hear if you have - and will try WindowSHAP in the meantime!

**Reply by matt-buckley** [reply]
@nickparative did you manage to use WindowSHAP with models with multivariate (or univariate) time series _outputs_ as well as inputs?

**Reply by St-Savin** [reply]
I have read the paper about the WindowSHAP method. While it seems the method function wells with the data used and seems more efficient, the method is not applied on simple systems to test how well it does vis-à-vis time series parameter like the delay, the gain, the time constant of a system and noise. The method seems focused on finding the most "interesting" part of a signal so the SHAP values are less diluted not finding any information about the system like the time delay between one variable and the output. I may be wrong, I only read the paper once. There is also the TimeSHAP, I only glanced at the paper, but it seems more focused on the characteristics of the system like the delay.

**Reply by St-Savin** [reply]
Just a little update, I made a dynamic system consisting of 2 first order systems (one has a delay the other not) (2 inputs), then I made a simple XGBoost model and analyze it with SHAP value, the results were good : the SHAP value estimated the dynamic response and the delay (Someone must have done this before,  but I wanted to code it myself ). Then, I modified the model so it would have a 2d input, but I encountered a lots of problems, it seems that I have to make a new model. So I should have something in 1-4 weeks for now.

**Reply by St-Savin** [reply]
So, I encountered far less problem than I expected. I made a GRU model that uses a 2d matrix and then analysed the model with SHAP values. I was able to obtain them by calling after  "tf.compat.v1.disable_v2_behavior()" after importing Tensorflow. Then I juggled a little the 3d array so I can use the summaryplot. I could not use the summaryplot on the 3d array of the SHAP values. I pretty sure that I was able the use it a few versions ago. I could make a jupyter notebook if you want, I don't mind I could use the practice since I never made one.

**Reply by St-Savin** [reply]
I have finished the analysis and found that there are ways to use SHAP values with a 2D input. I used "tf.compat.v1.disable_v2_behavior()" after "import tensorflow as tf". Someone else mentioned that it could be done with an earlier version of SHAP, but I did not try it. More importantly, I don't see why a neural network with 2D inputs should be used to model time series. The sampling frequency is determined by the fastest input, and the length of the window is determined by the slowest input. In short, it creates the biggest input for the model. Also, as it has been shown in the [repo](https://github.com/St-Savin/SHAP_times-series), a high sampling frequency will drown/diminish the SHAP values for the slowest variables.

**Reply by XiaoliangWang2001** [reply]
Hey @St-Savin , 

> Just a little update, I made a dynamic system consisting of 2 first order systems (one has a delay the other not) (2 inputs), then I made a simple XGBoost model and analyze it with SHAP value, the results were good : the SHAP value estimated the dynamic response and the delay (Someone must have done this before, but I wanted to code it myself ). Then, I modified the model so it would have a 2d input, but I encountered a lots of problems, it seems that I have to make a new model. So I should have something in 1-4 weeks for now.

Could you explain a bit how exactly did "the SHAP value estimated the dynamic response and the delay"? Were you using Shap or WindowSHAP or TimeSHAP?

**Comment by St-Savin** [comment]
This is the step response on u1 and u2 of the dynamic system that I created. Nothing special 2 input, 1 output. One of the input has a delay.
![image](https://github.com/user-attachments/assets/ec42ba6f-a4ff-47dc-9ae3-a0bc10441d27)
These are the performance of the differential model ( x1= x0 + K *delta u  ). The K were found with the SHAP values of an XGBoost model (purple and teal lines, the blue and orange one are form the mathematical model).
![image](https://github.com/user-attachments/assets/6d4502ee-6451-4b96-8310-a98dab346224)
So, these are great results, but it is a linear model. There is no need for SHAP values or XGBoost. I am planning to test the methodology with a non-linear model.
